### ISSUE TITLE
Randomize Framework order in runs

### ISSUE THREAD

**Issue by p8** [TechEmpower/FrameworkBenchmarks#8830]
Currently the benchmarks are run in the same order everytime.
Sometimes a run fails after a number of frameworks were benchmarked, or the run is restarted.
This causes the frameworks starting with `a` to have more test runs than frameworks starting with `z`.
If the order was randomized the number of runs would be better distributed.

**Comment by fakeshadow** [comment]
As someone maintaining a benchmark starts with `x` I feel this. 

That said IMO a fair way of handling order is to prioritize benches with most recent changes. As for benches that haven't changed for a while I guess the maintainers would care less about continuous run result.

**Comment by joanhey** [comment]
I think that is enough to start one run with `a` and the next with `z`.
And perhaps we still see some differences in the results.

This run and the next, change the servers, databases, ... so the change it's for all frameworks.
Don't depend from the changes in the frameworks.
A mature framework need less changes than a young one. Still we can bench in local to test small changes.

**Comment by itrofimow** [comment]
What I maintain starts with `u`, some I'm heavily biased here, but I would also appreciate this change being implemented.

My concern is not about failures or restarts, as they usually don't happen that often when the environment is stable, but rather about a feedback latency:
I mostly use TFB as a measurement tool (and a big shout-out to TE crew for providing that tool), and given a hypothetical performance drop in the ongoing run, I'm left with approx. a day to squeeze a potential fix into the next measurement, and a failure to do so would lead to a feedback latency of two full weeks (every run is approx. a week).
Moreover, any dependency bump I do is at least a week (an almost full run) in terms of feedback latency, and 1.5 weeks on average.

Flipping the order between runs (or FWIW randomizing it) would significantly reduce these latencies for me.

**Comment by joanhey** [comment]
The frameworks that stay in the middle have ~3 days to make changes.
It's the same if the bench begin with a or in reverse order.
The  problem is the frameworks that are the last in the run.
Please don't randomize, now we almost know when are the result for our framework.
But we need to flip the order in any new run !!

**Comment by NateBrady23** [comment]
I like the idea of flipping the order. I'm just getting back from vacation and catching up on a bunch of stuff. Let's get the environment stable, and then I think this is easy to do. Will leave this open until we get it in.

**Comment by joanhey** [comment]
After finish the last full run, the next run did not flip the order.

**Comment by volyrique** [comment]
That's because the [tfb-startup.sh](https://github.com/TechEmpower/FrameworkBenchmarks/blob/034d6fbec00ea745bfeabf5fde651e3d76196298/toolset/continuous/tfb-startup.sh#L6) script runs `tfb-shutdown.sh` on startup; the latter is responsible for flipping the order. Is changing the order only after an unsuccessful run by design?

**Comment by p8** [comment]
I think the following run was reversed: https://tfb-status.techempower.com/results/3c2e9871-9c2a-4ff3-bc31-620f65da4e74. The “last framework” tested is incorrect though.

**Comment by NateBrady23** [comment]
> That's because the [tfb-startup.sh](https://github.com/TechEmpower/FrameworkBenchmarks/blob/034d6fbec00ea745bfeabf5fde651e3d76196298/toolset/continuous/tfb-startup.sh#L6) script runs tfb-shutdown.sh on startup; the latter is responsible for flipping the order. Is changing the order only after an unsuccessful run by design?

No, I forgot that we actually run the shutdown script twice after a successful run because it's being called from the startup script as well. The design was supposed to be the exact opposite. I'll have to move it to the startup script and it will just reverse every time a run starts.

**Comment by volyrique** [comment]
@NateBrady23 It looks like now the opposite thing is happening - the order is always reversed, i.e. the implementations starting with `Z` run first.

**Comment by akupiec** [comment]
How about adding an option to run tests in the order of their last execution time, from fastest to slowest? 😈

**Comment by volyrique** [comment]
I am pretty sure that that approach would end up being effectively the same as running them in alphabetical order (or in random order at best), at the cost of significant implementation complexity.

**Comment by volyrique** [comment]
@msmith-techempower, after PR #10109 have the toolkit scripts been updated on the machine that is orchestrating the continuous benchmarking runs in the Citrine environment? Just cloning this repository, which is done by the `toolset/continuous/tfb-startup.sh` script, is not going to be enough, unless you have made some adjustments locally (other than those that are documented in the `toolset/continuous/tfb.service` file).

As far as I can tell, it should be sufficient to copy `toolset/continuous/tfb-startup.sh` to the `/home/techempower` directory and to set the script to be executable (again, assuming no local customizations). Furthermore, it shouldn't be necessary to interrupt the current continuous benchmarking run.

**Comment by msmith-techempower** [comment]
I must have hallucinated that copying those files out was part of our automated process - it is not.

I have just copied them from the latest to the home dir, they will be run next continuous cycle.

**Comment by joanhey** [comment]
Add every 4-5 runs to be like a cloud (not OK for some devs) but like a commodity server.
With docker we can run it with limited CPU cores, like 1 or 2, to simulate a cloud (or small server).
The number of CPU cores need to be small, but necessary to know the same FWs the results.

Before  the results show results for a big server and cloud servers also.

**Comment by volyrique** [comment]
Flipping the framework order after each continuous benchmarking run seems to work correctly now, so I suppose that this issue could be closed.
